<launch>
    <arg name="veh"/>
    <arg name="local" default="false" doc="true for running everything you possibly can on laptop. false for everything running on vehicle."/>
    <arg name="config" default="baseline" doc="Specify a config."/>
    <arg name="param_file_name" default="default" doc="Specify a param file. ex:megaman." />
    <arg name="camera" default="false" doc="camera set to false means only apriltags no camera nodes are launched"/>
    <arg name="live" default="true" doc="live set to false means we don't run the actual camera (i.e. running from a log file" />
    <arg name="apriltag" default="true" doc="apriltag set to false means no apriltag detection"/>
    <arg name="cam_info" default="true" doc="If set to true, use cam_info_reader_node to pusblish CameraInfo."/> 
    <arg name="raw" default="false" doc="If set to true, also publish raw image using decoder."/>
    <arg name="rect" default="false" />
    <arg name="mocap" default="false" />
    <arg name="laptop" default="false" />
    <arg name="duckiebot" default="false" />
    
    <group if="$(arg mocap)">
        <!-- camera_node -->
        <include  file="$(find pi_camera)/launch/camera_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="config" value="$(arg config)"/>
            <arg name="param_file_name" value="$(arg param_file_name)"/>
        </include>

        <!-- cam_info_reader_node -->
        <remap from="cam_info_reader_node/camera_info" to="camera_node/camera_info"/>
        <remap from="cam_info_reader_node/compressed_image" to="camera_node/image/compressed"/>
        <include file="$(find pi_camera)/launch/cam_info_reader_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
            <arg name="param_file_name" value="$(arg param_file_name)"/>
        </include>

        <!-- Waypoint Planning -->
    <remap from="mocap_waypoint_planning_node/vehicle_pose_pair" to="mocap_localization_node/vehicle_pose_pair"/>
        <include file="$(find mocap)/launch/mocap_waypoint_planning_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
            <arg name="param_file_name" value="$(arg param_file_name)"/>
        </include>
    </group>


    <group if="$(arg laptop)">
        <!-- decoder_node -->
        <remap from="decoder_node/compressed_image" to="camera_node/image/compressed"/>
        <remap from="decoder_node/image/raw" to="camera_node/image/raw"/>
        <include file="$(find pi_camera)/launch/decoder_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
            <arg name="param_file_name" value="$(arg param_file_name)"/>
        </include>

        <!-- camera raw image-->
        <remap from="raw_cam_info_reader_node/camera_info" to="camera_node/raw_camera_info"/>
        <remap from="raw_cam_info_reader_node/compressed_image" to="camera_node/image/raw"/>
        <include file="$(find pi_camera)/launch/cam_info_reader_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
            <arg name="param_file_name" value="$(arg param_file_name)"/>
            <arg name="node_name" value="raw_cam_info_reader_node" />
            <arg name="image_type" value="raw" />
        </include>

        <!-- Rectification -->
        <remap from="image_raw" to="camera_node/image/raw"/>
        <remap from="camera_info" to="camera_node/raw_camera_info"/>
        <remap from="image_rect_color" to="camera_node/image/rect" />
        <include file="$(find pi_camera)/launch/image_proc_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
        </include>

        <!-- AprilTags Detections -->
        <remap from="image_rect" to="camera_node/image/rect" />
        <remap from="camera_node/image/camera_info" to="camera_node/raw_camera_info" />
        <include if="$(arg apriltag)" file="$(find apriltags_ros)/launch/apriltag_detector_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
            <arg name="param_file_name" value="$(arg param_file_name)"/>
        </include>

        <!-- Location Estimation -->
        <remap from="mocap_localization_node/tag_detections" to="tag_detections"/>
        <include file="$(find mocap)/launch/mocap_localization_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
            <arg name="param_file_name" value="$(arg param_file_name)"/>
        </include>
     </group>

    <group if="$(arg duckiebot)">
        <include file="$(find duckietown)/machines"/>
        <!-- run inverse_kinematics_node -->
        <remap from="inverse_kinematics_node/car_cmd" to="/mocap1/mocap_waypoint_planning_node/car_cmd"/>
        <remap from="inverse_kinematics_node/wheels_cmd" to="wheels_driver_node/wheels_cmd" />
        <include file="$(find dagu_car)/launch/inverse_kinematics_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
        </include>

        <!-- run forward_kinematics_node -->
        <remap from="forward_kinematics_node/wheels_cmd" to="wheels_driver_node/wheels_cmd_executed" />
        <include file="$(find dagu_car)/launch/forward_kinematics_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
        </include>

        <!-- run the velocity_to_pose_node -->
        <remap from="velocity_to_pose_node/velocity" to="forward_kinematics_node/velocity" />
        <include file="$(find dagu_car)/launch/velocity_to_pose_node.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="local" value="$(arg local)"/>
            <arg name="config" value="$(arg config)"/>
        </include>

        <!-- Wheels Driver -->
        <include file="$(find dagu_car)/launch/wheels_driver_node.launch">
            <arg name="veh" value="$(arg veh)"/>
        </include>
    </group>


</launch>
